### ISSUE TITLE
Agent in trouble, but does not exit - tar: Error is not recoverable: exiting now

### ISSUE THREAD

**Issue by feliksik** [rancher/rancher#1370]
I am starting an agent with (AFAIK) the correct server key. The agent stays up, but does not appear in the server. 
Logs give below message, 165 times. Note that I did install rancher-agent on this host before, and removed/purged it from the server. (maybe the server is not clearly resetting it's state?)

It would be nice to have better diagnostics on what went wrong etc. 
In general, it would be good to have some simple way of checking (like exit status, of maybe some rest call to the agent) whether the agent permanently failed with connecting to the server (e.g. if you want to automate the installation of rancher agent on a vm)

```
INFO: Starting agent for 17D99798E3E9D0EEA8ED
INFO: Access Key: 17D99798E3E9D0EEA8ED
INFO: Config URL: http://demo.domain.com/v1
INFO: Storage URL: http://demo.domain.com/v1
INFO: API URL: http://demo.domain.com/v1
INFO: IP: 130.211.91.22
INFO: Port:
INFO: Required Image: rancher/agent:v0.7.9
INFO: Current Image: rancher/agent:v0.7.9
INFO: Using image rancher/agent:v0.7.9
INFO: Downloading agent http://demo.domain.com/v1/configcontent/configscripts{"id":"2f146fc7-4a49-46fb-a5c9-d500f3fb75fe","type":"error","links":{},"actions":{},"status":401,"code":"Unauthorized","message":"Unauthorized","detail":null}gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
INFO: Starting agent for 17D99798E3E9D0EEA8ED
INFO: Access Key: 17D99798E3E9D0EEA8ED
INFO: Config URL: http://demo.domain.com/v1
INFO: Storage URL: http://demo.domain.com/v1
INFO: API URL: http://demo.domain.com/v1
INFO: IP: 130.211.91.22
INFO: Port:
INFO: Required Image: rancher/agent:v0.7.9
INFO: Current Image: rancher/agent:v0.7.9
INFO: Using image rancher/agent:v0.7.9
INFO: Downloading agent http://demo.domain.com/v1/configcontent/configscripts{"id":"3bfd75de-3aa3-42d1-bc32-18ec29d88f28","type":"error","links":{},"actions":{},"status":401,"code":"Unauthorized","message":"Unauthorized","detail":null}gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now


```

**Comment by cloudnautique** [comment]
The agent is trying to get the configuration information that it needs, like python-agent, host-api, and other package versions/config. It looks like the token its using is not able to auth to the server.

 Any chance this is a new rancher environment or Auth settings changed? Is this host being reused/provisioned? If so, have you tried stopping the agents, removing rancher-agent-state container then try reregistering?

**Comment by cloudnautique** [comment]
Sorry, your first line answered most of my questions. There is a rancher-agent-state container that might be interfering with the auth process. If you can remove the rancher/agent instances and volumes then re-register it might clear things up.

**Comment by ebr** [comment]
+1 - in the same situation right now. A host was previously added to the server; I removed the agent from the host, and purged it from the server. now I'm unable to join this host to the server again. The error messaging is identical to @feliksik's

**Comment by aburnett** [comment]
+1
401 from the rancher server when trying to re-add a host with the same name.

**Comment by cloudnautique** [comment]
@ebr @aburnett are there files in /var/lib/rancher/state when you do `ls -a` ? If so, or just for good measure :), remove the directory. That is where Rancher stores its actual state information.

**Comment by aburnett** [comment]
@cloudnautique on the server or agent hosts? Sadly I just blew away my agent VMs and created new ones (with new names) to get unblocked. I'll try and replicate w/ my new VMs later on.

**Comment by cloudnautique** [comment]
On the agent hosts. When an agent starts it gathers info, creates a rancher-agent-state container then launches rancher-agent. That state container creates a bind mount to /var/lib/rancher/state on the host so that it can persist data.

**Comment by sebmoule** [comment]
+1 I get exactly the same problem, ant thanks to @cloudnautique when I delete the ```/var/lib/rancher/state``` directory on my host, I was able again to register.

**Comment by sdlarsen** [comment]
I experienced the exact same issue. My fix was a little different. I actually have files in /var/lib/rancher/state from before the fix, so deleting them might have solved it.
I ended up deleting the environment in rancher and recreate it. Then after purging the rancher containers/images from the machines (again) register them with the new auth key from the new environment. I used identical environment name and host names.

**Comment by andyshinn** [comment]
I just ran into this same issue. I moved my Rancher server to a new host (preserving the MySQL data). I then deleting the agent from the UI, deleting the agent containers (agent, instance, and network), then started the new agent based on the config from the UI custom host docker run.

I removed the /var/lib/rancher/state folder and was able to start it up. It would be nice to have the ability to migrate hosts in a reasonable way at least. I posed the question around migration at http://forums.rancher.com/t/how-to-reconnect-agents-to-moved-server/928.

**Comment by flaccid** [comment]
I also run into this with use case of re-using hosts (actually more common than you think, particularly in static servers of corporates). Also used removal of `/var/lib/rancher/state` as workaround.

**Comment by demarant** [comment]
I support both @flaccid and @andyshinn suggestions. It is also for us, very common to unregister and move custom host from one environment to another, e.g. from devel to staging during the development stage of new applications. Also common when you kind of need to change the way the host is registered, e.g. adding new persistent labels or specific docker-engine settings ... so indeed there should be an easier way to move host from one environment to another or re-register without the need to ssh into the host a second time and manually delete /var/lib/rancher/state folder.

**Comment by npcode** [comment]
I couldn't add a host if access control is enabled because of this error.

**Comment by deniseschannon** [comment]
@npcode You should try removing `/var/lib/rancher/state` before adding an agent if you hit this error. Are you re-using an agent?

**Comment by npcode** [comment]
@deniseschannon It works after removing the directory. Thanks a lot!

**Comment by jvdgeest** [comment]
Thanks @deniseschannon, that solved the issue for me.

**Comment by llitfkitfk** [comment]
@deniseschannon thx

Docker for Mac:
if someone using docker for mac
try to run container: docker run --rm -it  -v /var/lib:/var/lib  alpine sh
then to rm -rf /var/lib/rancher/state

**Comment by defunctzombie** [comment]
Will this be fixed in 1.2.0? It is non-obvious when this happens how to fix it since removing the agent and agent state containers alone doesn't do it.

**Comment by deniseschannon** [comment]
@defunctzombie We are actually introducing a new agent (written in Go instead of Python) in 1.2.0. 

@StrongMonkey Can you try out moving a host from one environment to another to see if you can reproduce these issues? If not, we will close this issue and if we see these issues after 1.2.0, we can open a new issue.

**Comment by StrongMonkey** [comment]
@deniseschannon  Same as https://github.com/rancher/rancher/issues/4073. I can try to make PR to script to see if that solve the probelm.

**Comment by deniseschannon** [comment]
Merged after v1.2.0-pre4-rc5

**Comment by sangeethah** [comment]
@ibuildthecloud , I tried to test the following scenario and it works fine even when testing with v1.2.0-pre4-rc5:

Added a custom host by running registration url command on an Unbuntu VM.
Host gets to "Active" state successfully.
Deployed few containers on this host.
Now "Deactive" and "Delete" this host.

From the host:
docker rm -f  <rancher/agent-instance> <rancher/agent>
Rerun registration url command.
Host gets registered successfully and gets to "Active" state.

Is the fix provided to address this scenario where user need not have to delete the stale  /var/lib/rancher/state before having to reregister the same host ? It seems to be working even before 
https://github.com/rancher/cattle/issues/2117 .

@StrongMonkey , Does this PR address this scenario already?  -  https://github.com/rancher/rancher/issues/6346

**Comment by sangeethah** [comment]
@ibuildthecloud @StrongMonkey , I am also able to delete a host from from 1 environment and add it to another environment successfully  using  v1.2.0-pre4-rc5.

Please confirm , what needs to done to specifically test rancher/cattle#2117 ?

**Comment by deniseschannon** [comment]
@sangeethah I don't think there's any specific use case to test it, but there have been times where people have had issues. Closing as you don't seem to face the issue.
